osu.Game/Beatmaps/BeatmapPanelBackgroundTextureLoaderStore.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using osu.Framework.Graphics.Textures;
+using osu.Framework.IO.Stores;
+using SixLabors.ImageSharp;
+using SixLabors.ImageSharp.Processing;
+
+namespace osu.Game.Beatmaps
+{
+    // Implementation of this class is based off of `MaxDimensionLimitedTextureLoaderStore`.
+    // If issues are found it's worth checking to make sure similar issues exist there.
+    public class BeatmapPanelBackgroundTextureLoaderStore : IResourceStore<TextureUpload>
+    {
+        // These numbers are taken from the draw visualiser size requirements for song select panel textures at extreme aspect ratios.
+        private const int height = 130;
+        private const int max_width = 1280;
+
+        private readonly IResourceStore<TextureUpload>? textureStore;
+
+        public BeatmapPanelBackgroundTextureLoaderStore(IResourceStore<TextureUpload>? textureStore)
+        {
+            this.textureStore = textureStore;
+        }
+
+        public void Dispose()
+        {
+            textureStore?.Dispose();
+        }
+
+        public TextureUpload Get(string name)
+        {
+            var textureUpload = textureStore?.Get(name);
+
+            // NRT not enabled on framework side classes (IResourceStore / TextureLoaderStore), welp.
+            if (textureUpload == null)
+                return null!;
+
+            return limitTextureUploadSize(textureUpload);
+        }
+
+        public async Task<TextureUpload> GetAsync(string name, CancellationToken cancellationToken = new CancellationToken())
+        {
+            // NRT not enabled on framework side classes (IResourceStore / TextureLoaderStore), welp.
+            if (textureStore == null)
+                return null!;
+
+            var textureUpload = await textureStore.GetAsync(name, cancellationToken).ConfigureAwait(false);
+
+            if (textureUpload == null)
+                return null!;
+
+            return await Task.Run(() => limitTextureUploadSize(textureUpload), cancellationToken).ConfigureAwait(false);
+        }
+
+        private TextureUpload limitTextureUploadSize(TextureUpload textureUpload)
+        {
+            var image = Image.LoadPixelData(textureUpload.Data.ToArray(), textureUpload.Width, textureUpload.Height);
+
+            // The original texture upload will no longer be returned or used.
+            textureUpload.Dispose();
+
+            Size size = image.Size();
+            int usableWidth = Math.Min(max_width, size.Width);
+
+            // Crop the centre region of the background for now.
+            Rectangle cropRectangle = new Rectangle(
+                (size.Width - usableWidth) / 2,
+                size.Height / 2 - height / 2,
```

Comment:
This implicitly assumes that the background will be taller than `height`. Which is not an entirely unreasonable assumption, but *if* the background isn't taller than 130px, this will get thrown:

```
[runtime] 2023-06-08 11:44:39 [error]: Background failed to load
[runtime] 2023-06-08 11:44:39 [error]: System.ArgumentException: Crop rectangle should be smaller than the source bounds. (Parameter 'cropRectangle')
[runtime] 2023-06-08 11:44:39 [error]: at SixLabors.ThrowHelper.ThrowArgumentException(String message, String name)
[runtime] 2023-06-08 11:44:39 [error]: at SixLabors.ImageSharp.Processing.Processors.Transforms.CropProcessor..ctor(Rectangle cropRectangle, Size sourceSize)
[runtime] 2023-06-08 11:44:39 [error]: at SixLabors.ImageSharp.Processing.CropExtensions.Crop(IImageProcessingContext source, Rectangle cropRectangle)
[runtime] 2023-06-08 11:44:39 [error]: at osu.Game.Beatmaps.BeatmapPanelBackgroundTextureLoaderStore.<>c__DisplayClass7_0.<limitTextureUploadSize>b__0(IImageProcessingContext i) in /home/dachb/Documents/opensource/osu/osu.Game/Beatmaps/BeatmapPanelBackgroundTextureLoaderStore.cs:line 79
```

with an error notification and all. Probably something we'll want to guard against given it shouldn't be much work at all.